src/main/java/hudson/plugins/gradle/injection/MavenInjectionEnvironmentContributor.java

```diff
@@ -0,0 +1,105 @@
+package hudson.plugins.gradle.injection;
+
+import hudson.EnvVars;
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.model.Computer;
+import hudson.model.EnvironmentContributor;
+import hudson.model.Executor;
+import hudson.model.Node;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static hudson.plugins.gradle.injection.MavenOptsHandler.MAVEN_OPTS;
+
+@Extension
+public class MavenInjectionEnvironmentContributor extends EnvironmentContributor implements MavenInjectionAware {
+
+    private static final MavenOptsHandler MAVEN_OPTS_HANDLER = new MavenOptsHandler(
+            MAVEN_EXT_CLASS_PATH_PROPERTY_KEY,
+            BUILD_SCAN_UPLOAD_IN_BACKGROUND_PROPERTY_KEY,
+            GRADLE_ENTERPRISE_ALLOW_UNTRUSTED_SERVER_PROPERTY_KEY,
+            GRADLE_ENTERPRISE_URL_PROPERTY_KEY
+    );
+
+    private final MavenExtensionsHandler extensionsHandler = new MavenExtensionsHandler();
+
+    @Override
+    public void buildEnvironmentFor(@Nonnull Run run, @Nonnull EnvVars envs, @Nonnull TaskListener listener) {
+        Executor executor = run.getExecutor();
+        if (executor == null) {
+            return;
+        }
+
+        Node node = executor.getOwner().getNode();
+        if (node == null) {
+            return;
+        }
+
+        FilePath nodeRootPath = node.getRootPath();
+        if (nodeRootPath == null) {
+            return;
+        }
+
+        if (!isInjectionEnabled(node)) {
+            return;
+        }
+
+        InjectionConfig config = InjectionConfig.get();
+        String server = config.getServer();
+
+        List<FilePath> extensions = new LinkedList<>();
+        extensions.add(extensionsHandler.getExtensionLocation(MavenExtensionsHandler.MavenExtension.GRADLE_ENTERPRISE, nodeRootPath));
```

Comment:
I'm a bit worried by the cost of calling `getExtensionLocation` since it's doing IO (maybe remote IO ?) and this will be called many times. Maybe let's be on the safe side and follow the Jenkins javadoc:

> This method can be called repeatedly for the same Run, thus the computation of this method needs to be efficient. If you have a time-consuming computation, one strategy is to take the hit once and then add the result as InvisibleAction to Run

Reply:
Are you talking about the `FilePath.child(...)` call underneath?